[EGM HLT@Phase2] Phase out caloTowers

swagata87 opened the pull request:
#### PR description:

With this PR we propose to discontinue the usage of caloTowers in Phase2 E/gamma HLT.  
This is done in a similar way as it was done for Run3 EGM HLT as described in [1], [2], [3], [4].

[1] https://its.cern.ch/jira/browse/CMSHLT-2286
[2] https://indico.cern.ch/event/1116957/contributions/4716661/attachments/2383419/4072743/caloTowerEGMHLT.pdf
[3] https://github.com/cms-sw/cmssw/pull/35049
[4] https://github.com/cms-sw/cmssw/pull/36157

#### PR validation:

RecHit-based H/E and CaloTower-based H/E are very close. This plot is made from `CMSSW_12_4_8`, using `ZprimeToEE` sample (`PhaseIISpring22`, `PU200`). The objects entering the plot are all the HLT-level superclusters with pT>10 GeV, in barrel. 
<img width="500" alt="Screen Shot 2023-01-13 at 23 15 11" src="https://user-images.githubusercontent.com/5052706/212429331-d062f1b3-ce11-4711-8441-c27432953141.png">

Tagging HLT upgrade convenors @beaucero and @SohamBhattacharya 

Backport of this PR is not necessary.

swagata87 committed:
phase out calotowers

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40525/33732

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @swagata87 (Swagata Mukherjee) for master.

It involves the following packages:

- HLTrigger/Configuration (**hlt**)


@cmsbuild, @missirol, @Martin-Grunewald can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @silviodonato this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented on HLTrigger/Configuration/python/HLT_75e33/modules/hltFixedGridRhoFastjetAllCaloForMuons_cfi.py:
```diff
@@ -1,7 +1,13 @@
+    eeRecHitsTag = cms.InputTag( 'hltEcalRecHit','EcalRecHitsEE' ),
+    skipHCAL = cms.bool( False ),
+    skipECAL = cms.bool( False ),
+    eThresHB = cms.vdouble( 0.1, 0.2, 0.3, 0.3 ),
+    eThresHE = cms.vdouble( 0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2 )
```
These look like the Run-3 thresholds. Are they the same for Phase 2 ?

swagata87 replied:
Let's double check this with @cms-sw/hcal-dpg-l2 and @cms-sw/pf-l2 
Is the following still valid for phase2? Or do we have new PF recHit threshold for HB for phase2?
https://github.com/cms-sw/cmssw/blob/33ef05739ed403d4af6ba915a406905080dbc5ca/RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHBHE_cfi.py#L5

missirol replied:
I thought this depends on the 'aging' scenario, e.g.
https://github.com/cms-sw/cmssw/blob/af4b7abc8eb5bf1bcf2dd39e576e76ea1bd0b464/SLHCUpgradeSimulations/Configuration/python/aging.py#L96
, and it would match what is used in the modules below [*], where there are already some differences.

Indeed, it would be good to clarify this.

[*]

https://github.com/cms-sw/cmssw/blob/af4b7abc8eb5bf1bcf2dd39e576e76ea1bd0b464/HLTrigger/Configuration/python/HLT_75e33/modules/particleFlowRecHitHBHE_cfi.py#L16
https://github.com/cms-sw/cmssw/blob/af4b7abc8eb5bf1bcf2dd39e576e76ea1bd0b464/HLTrigger/Configuration/python/HLT_75e33/modules/particleFlowClusterHBHE_cfi.py#L13

https://github.com/cms-sw/cmssw/blob/af4b7abc8eb5bf1bcf2dd39e576e76ea1bd0b464/HLTrigger/Configuration/python/HLT_75e33/modules/hltParticleFlowRecHitHBHEForEgamma_cfi.py#L25
https://github.com/cms-sw/cmssw/blob/af4b7abc8eb5bf1bcf2dd39e576e76ea1bd0b464/HLTrigger/Configuration/python/HLT_75e33/modules/hltParticleFlowClusterHBHEForEgamma_cfi.py#L13

abdoulline replied:
I was about to point to usual Phase2 aging customization
https://github.com/cms-sw/cmssw/blob/master/SLHCUpgradeSimulations/Configuration/python/aging.py#245 
With PFlowClusters thresholds/seeds involved.
But just noticed that Marino has already commented on it _several minutes ago_...

missirol reviewed (commented):
I'm a bit confused.

Doesn't the change to `hltFixedGridRhoFastjetAllCaloForMuons` also affect the muon reconstruction?

If that is intended, and the same module is used by both EGM and MUO, the postfix "ForMuons" could be removed for clarity.

Several EGM modules (e.g. `hltDiEG2312IsoHEL1SeededFilter`) still reference `hltFixedGridRhoFastjetAllCaloForEGamma` (which doesn't exist); I guess it does not matter because `doRhoCorrection = False`, but this should be replaced with an empty InputTag (like in `hltDiEG2312IsoClusterShapeL1SeededFilter`), or the name of an appropriate (existing) collection (e.g. `hltFixedGridRhoFastjetAllCalo(ForMuons)`).

The files `hltTowerMakerForAllForEgamma_cfi` and `hltRegionalTowerForEgamma_cfi` should be removed if not used (unless there is a good reason to keep them).

swagata87 committed:
addressed Marino's comments

swagata87 commented:
> Doesn't the change to `hltFixedGridRhoFastjetAllCaloForMuons` also affect the muon reconstruction?

Indeed it does. But for now RhoCorrection is disabled in muon HLT also, so no difference will show up in tests. 

> If that is intended, and the same module is used by both EGM and MUO, the postfix "ForMuons" could be removed for clarity.

I have now kept `hltFixedGridRhoFastjetAllCaloForMuons` as it is and created a new file `hltFixedGridRhoFastjetAllCaloForEGamma`. This way, later on MUO POG can configure the `hltFixedGridRhoFastjetAllCaloForMuons` differently than egamma if they need to. 
 
> Several EGM modules (e.g. `hltDiEG2312IsoHEL1SeededFilter`) still reference `hltFixedGridRhoFastjetAllCaloForEGamma` (which doesn't exist); I guess it does not matter because `doRhoCorrection = False`, but this should be replaced with an empty InputTag (like in `hltDiEG2312IsoClusterShapeL1SeededFilter`), or the name of an appropriate (existing) collection (e.g. `hltFixedGridRhoFastjetAllCalo(ForMuons)`).

this issue is now solved as I have now introduced a new module named `hltFixedGridRhoFastjetAllCaloForEGamma`

> The files `hltTowerMakerForAllForEgamma_cfi` and `hltRegionalTowerForEgamma_cfi` should be removed if not used (unless there is a good reason to keep them).

these 2 files are now removed

The latest changes are tested with `runTheMatrix.py -l 20834.76`

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40525/33733

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #40525 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

swagata87 committed:
use correct thresholds for HCAL

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40525/33736

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #40525 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
please test

>  If it's fine, it can be done in a followup PR.

That'd be okay with me.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdd063/30000/summary.html
**COMMIT**: fda2b1064ae1191e3bb06985585699ee6f501ed8
**CMSSW**: CMSSW_13_0_X_2023-01-15-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40525/30000/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555510
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-15-2300+cdd063/55020/triggerResults): no differences found

missirol commented:
+hlt

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
